kotlin/authenticate-with-oauth/src/main/java/com/esri/arcgisruntime/sample/authenticatewithoauth/MainActivity.kt

```diff
@@ -0,0 +1,343 @@
+      } catch (e: Exception) {
+        // auth code has likely expired, clear existing auth code and begin OAuth flow
+        getString(R.string.error_auth_exception, e.message).let {
+          Log.d(TAG, it)
+          runOnUiThread {
+            logToUser(it)
+          }
+        }
+        sharedPreferences.clearAuthCode()
+        beginOAuth()
+      }
+    }
+    return AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CANCEL, null)
+  }
+
+  private fun beginOAuth() {
+    // get the authorization code by sending user to the authorization screen
+    val authorizationUrl = OAuthTokenCredentialRequest.getAuthorizationUrl(
+      oAuthConfig.portalUrl, oAuthConfig.clientId, oAuthConfig.redirectUri, ACCESS_TOKEN_EXPIRY_MINS
+    )
+
+    // create an Intent to attempt to handle the authorization URL
+    with(Intent(Intent.ACTION_VIEW, Uri.parse(authorizationUrl))) {
+      this.resolveActivity(packageManager)?.let {
+        // this identifier ensures that the browser will attempt to reuse the same window each time the application
+        // launches the browser with the same identifier, which in this case is static and will always only be our
+        // application ID
+        this.putExtra(Browser.EXTRA_APPLICATION_ID, BuildConfig.APPLICATION_ID)
+        startActivity(this)
+        return
+      }
+    }
+
+    // user doesn't have a browser available to handle the Intent so we use a WebView to handle OAuth. WebView methods
+    // must be called on UI thread
+    runOnUiThread {
+      setupWebView()
+      webView.loadUrl(authorizationUrl)
+    }
+  }
+
+  private fun setupWebView() {
+    // setup a WebViewClient to override handling of custom scheme and host for Intent
+    webView.webViewClient = object : WebViewClient() {
+      override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
+        Uri.parse(url)?.let {
+          if (it.scheme == "my-ags-app" && it.host == "auth") {
+            startActivity(generateAuthIntent(it))
+            return true
+          }
+        }
+        return super.shouldOverrideUrlLoading(view, url)
+      }
+
+      @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
+      override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
+        if (request?.url?.scheme == "my-ags-app" && request.url?.host == "auth") {
+          startActivity(generateAuthIntent(request.url))
+          return true
+        }
+        return super.shouldOverrideUrlLoading(view, request)
+      }
+    }
+
+    // enabled to allow javascript to run on auth webpage.
+    webView.settings.javaScriptEnabled = true
+    webView.visibility = View.VISIBLE
+  }
+
+  /**
+   * Generate the Intent that launches the MainActivity with the new auth code
+   *
+   * @param uri instance of Uri generated from URL that OAuth webpage redirects to after successful login
+   */
+  private fun generateAuthIntent(uri: Uri): Intent {
```

Comment:
I guess its this function here where we need to generate a unique intent flag to not accidently call other similarly IDed Activities?

Reply:
What we define in the manifest must be passed to the Portal as part of the request to Auth.
```
 <intent-filter>
                <action android:name="android.intent.action.VIEW" />

                <category android:name="android.intent.category.DEFAULT" />
                <category android:name="android.intent.category.BROWSABLE" />

                <data
                    android:host="auth"
                    android:scheme="my-ags-app" />
            </intent-filter>
```

`my-ags-app://auth`

So what we do is when we're creating our `OAuthConfiguration`, we pass `my-ags-app://auth` as the re-direct URI so when a user auths successfully, the user is redirected to that URI which our app has declared as it can handle in it's manifest.

If we changed the URI here, which we could do but we'd need to add another `intent-filter` to the manifest, we'd still be redirected to `my-ags-app://auth` as the portal is forcing this:
https://github.com/Esri/arcgis-runtime-samples-android/pull/590#issuecomment-473006032